### DISCUSSION TITLE
How to transfer PROP and ADD error to mrgsolve  if estimated as fixed effects?

### DISCUSSION THREAD

**Discussion by Anks2030** [metrumresearchgroup/mrgsolve#959] [Q&A]
Hi @kylebaron 
I have a model which has PROP and ADD error estimated as fixed effects parameters.
Do you know how to code ADD and PROP errors to mrgsolve if they are not estimated as variances under $SIGMA?

My mrgsolve code for the 2CM model with FO abs and elimination is below. Kindly suggest any errors in coding. 
I will highly appreciate it if you can pass on any previous discussions in the group or if you can quickly look into it

$PROB 

$PARAM TVV2 = 2640, TVCL = 121, TVV3 = 2200, TVQ =317, TVKA = 0.15, TVF1 = 0.656, WT=70, PROP=0.15, ADD=0.001

$PKMODEL cmt="GUT CENT PERIPH", depot=TRUE

$SET end=240, delta=0.5

$MAIN
double V2 = exp(log(TVV2) + 0.835*log(WT/70) + EV2);
double CL = exp(log(TVCL) + 0.759*log(WT/70) + ECL);
double V3 = exp(log(TVV3)  + EV3);
double Q  = exp(log(TVQ)  + EQ);
double KA = exp(log(TVKA));
F_CENT    = exp(log(TVF1)  + EF);

$OMEGA @labels EV2 ECL EV3 EQ EF
0.318 0.087 0.235 0.214 0.216 

$SIGMA @labels EPS1 EPS2
1 1

$TABLE
capture IPRED = CENT/V2;
capture DV = IPRED*(1+PROP*EPS(1))+ADD*EPS(2);

$CAPTURE IPRED CL V2 DV

My NM code is below
$SUBROUTINE ADVAN4 TRANS4

$PK 

TVV2  = THETA(1)		                      ; TYPICAL VALUE V1
V2    = TVV2*EXP(ETA(1))*(WT/70)**THETA(8)            ; CENTRAL CMT VOL
TVCL2 = THETA(2)		                      ; TYPICAL VALUE CLP
CL    = TVCL2*EXP(ETA(2))*(WT/70)**THETA(7)           ; CL
V3    = THETA(3)*EXP(ETA(3))                ; PERIPHERAL CMT VOL 
Q     = THETA(4)*EXP(ETA(4))                ; DISTRIBUTION
KA    = THETA(5)                            ; SC ABS RATE 
F1    = THETA(6)*EXP(ETA(5))                ; BIOAVAILABILITY

$ERROR
CP = A(2)/V2
IPRED = CP                          ; PK
Y = IPRED*(1+THETA(9)*EPS(1))+THETA(10)*EPS(2)  
IRES=DV-IPRED                                    
W = IPRED                                                     ; WEIGHTING FOR PROPORTIONAL ERROR 
IWRES=IRES/(W+.0001)                  
        

$THETA
2640        ; 1 V2 (mL)   
121         ; 2 CLP (mL/day)  
2200        ; 3 V3 (mL)  
317         ; 4 Q (mL/day)  
0.15        ; 5 KA (1/day) 
0.656       ; 6 Bioavailability
0.759       ; 7 BW on CL
0.835       ; 8 BW on V2 
0.15        ; 9 proportional error
0.001       ; 10 additive error

$OMEGA  
0.318       ; 1 IIV ON V2
0.087       ; 2 IIV ON CL
0.235       ; 3 IIV ON V3
0.214       ; 4 IIV ON Q
0.216       ; 5 IIV ON F

$SIGMA  
1 FIX       ; EPS1 
1 FIX       ; EPS2

Best regards,
-Ankur

**Comment by kylebaron** [comment]
Hi @Anks2030  - 

You code it just the same way as you did in nonmem model; You can still reference `EPS(1)` and `EPS(2)` and write 

If this was the nonmem model

```r
$ERROR
CP = A(2)/V2
IPRED = CP ; PK
Y = IPRED*(1+THETA(9)*EPS(1))+THETA(10)*EPS(2)
```

Then you'd write 
```r
$PROB

$PARAM TVV2 = 2640, TVCL = 121, TVV3 = 2200, TVQ =317, TVKA = 0.15, TVF1 = 0.656, WT=70, PROP=0.15, ADD=0.001

$SIGMA @labels EPS1 EPS2
1 1

$TABLE
capture IPRED = CENT/V2;
capture DV = IPRED*(1+PROP*EPS(1))+ADD*EPS(2);
```

Or given you have provided labels

```r
$SIGMA @labels EPS1 EPS2
1 1

$TABLE
capture IPRED = CENT/V2;
capture DV = IPRED*(1+PROP*EPS1)+ADD*EPS2;
```

I think some punctuation got dropped when you pasted the code here but it looks like it's right.

**Reply by prewr1** [reply]
If the combined model is coded in NONMEM as W=sqrt((F*PROP)^2+ADD^2), Y=IPRED+W*EPS(1) and $SIGMA 1 FIX. Then in this case how is this implemented in mrgsolve?

**Reply by kylebaron** [reply]
It's the same way you implement in NONMEM; `PROB` and `ADD` could come from a variety of places, just coding it here as a "parameter"

```
$PARAM PROP = 0.02, ADD = 10

$TABLE
double W = sqrt(pow(F*PROP, 2) + pow(ADD,2));
double Y = IPRED + W * EPS(1); 

$SIGMA 1
```

**Comment by kylebaron** [comment]
@Anks2030 BTW ... I converted this to a discussion; you can keep the conversation going here:
https://github.com/metrumresearchgroup/mrgsolve/discussions/959

**Reply by Anks2030** [reply]
Thanks @kylebaron 
So basically how I coded PROP and ADD in the above mrgsolve are correct?
I just have to pick the variances from the NM  .LST file to update the $OMEGA and $SIGMA?
Thanks for your quick response. Best regards,


Another suggestion regarding coding in mrgsolve for $MAIN

# In Nonmem this is how I coded V2
V2    = TVV2 * (WT/70)**THETA(8) * EXP(ETA(1))

# In mrgsolve I coded this way
$MAIN
double V2 = exp(log(TVV2) + 0.835*log(WT/70) + EV2);

What is the benefit of coding in the above way as compared to the below code?

$MAIN
double V2 = V2 * pow(WT/70, 0.835)*EXP(EV2)

Thanks and Regards

-Ankur

**Reply by smouksassi** [reply]
just one comment about 
double V2 = exp(log(TVV2) + 0.835*log(WT/70) + EV2);
this parametrization ensures positive parameters even when you simulate with uncertainty and is preferred whenever you are simulating with parametric multivariate normal / student uncertainty.
but as Kyle mentions try to write the model like this in NONMEM and fit in this format before going to mrgsolve.
mrgsolve just write the model and translate it as it was written otherwise mistakes happen

**Comment by kylebaron** [comment]
Hi @Anks2030  - 

I would encourage you to write the model in mrgsolve with the same _format_ that you used in NONMEM; some of the syntax will be slightly different, but keeping the same format will help you avoid mistakes in translating the model. So I'd prefer the second proposal because it's easier to keep in sync with your NONMEM model.

Thanks,
Kyle

<hr>

P.S. 

You don't have to use these, but I think it's worth asking about: 

- there is a code block called `$NMEXT` that will import all the nonmem estimates for you; you can read more about it [here](https://mrgsolve.org/user_guide/model-specification.html?q=nmext#block-nmext)
- there is also a plugin that let's you use more nonmem-like syntax; you can read more about that [here](https://mrgsolve.org/user_guide/plugins.html?q=nm-vars#plugin-nm-vars)

You might have some other purposes with this model making these inappropriate for your case; but if the mrgsolve model is getting developed and used alongside the NONMEM model, I'd highly encourage taking a look.

**Reply by Anks2030** [reply]
Thanks, @kylebaron for the quick responses and suggestions. Regards!
